dwave/system/temperatures.py

```diff
@@ -0,0 +1,488 @@
+    for sI in range(samples.shape[0]):
+        np.add.at(effective_fields[sI,:],irow,qdata*samples[sI,icol])
+        np.add.at(effective_fields[sI,:],icol,qdata*samples[sI,irow])
+
+    if current_state_energy is True:
+        #Ising: eff_field = 2*s*(h + J*s)
+        effective_fields = 2*samples*effective_fields
+        
+    return effective_fields
+
+def maximum_pseudolikelihood_temperature(bqm = None,
+                                         sampleset = None,
+                                         site_energy = None,
+                                         bootstrap_size = 0,
+                                         seed=None,
+                                         Tguess=None):
+    '''Returns a sampling-based temperature estimate.
+    
+    The temperature T parameterizes the Boltzmann distribution as 
+    P(x) = exp(-H(x)/T)/Z(T), where P(x) is a probability over a state space, 
+    H(x) is the energy function (BQM) and Z(T) is a normalization. 
+    Given a sample set a temperature estimate establishes the temperature that 
+    is most likely to have produced the sample set.
+    An effective temperature can be derived from a sampleset by considering the 
+    rate of excitations only. A maximum-pseudo-likelihood (MPL) estimator 
+    considers local excitations only, which are sufficient to establish a 
+    temperature efficiently (in compute time and number of samples). If the bqm
+    consists of independent variable problems (no couplings between variables) 
+    then the estimator is equivalent to a maximum likelihood estimator.
+    
+    The effective MPL temperature is defined by the solution T to 
+    0 = sum_i \sum_{s in S}} nu_i(s) exp(nu_i(s)/T), 
+    where nu is the energy lost in flipping spin i against its current 
+    assignment (the effective field). 
+    
+    The problem is a convex root solving problem, and is solved with scipy 
+    optimize.
+
+    If the distribution is not Boltzmann with respect to the BQM provided, as
+    may be the case for heuristic samplers (such as annealers). The temperature
+    estimate can be interpretted as characterizing only a rate of local 
+    excitations. In the case of sample sets obtained from D-Wave quantum 
+    annealing the temperature can be identified with a physical temperature via
+    a late-anneal freeze-out phenomena.
+    
+    Args:
+        bqm (:obj:`dimod.BinaryQuadraticModel`, optional):
+            Binary quadratic model describing sample distribution.
+            If bqm and site_energy are both None, then by default 
+            100 samples are drawn from the default DWaveSampler(), 
+            with bqm defaulted as described.
+        sampleset (:class:`dimod.SampleSet`, optional):
+            A set of samples, assumed to be fairly sampled from
+            a Boltzmann distribution characterized by bqm.
+        site_energy (2d numpy array, optional):
+            The effective fields associated to a sampleset and bqm. Rows
```

Comment:
fields associated with ``sampleset`` and ``bqm`` (or "with the specified sample set and BQM")